File: src/main/java/org/jukeboxmc/item/enchantment/Enchantment.java
package org.jukeboxmc.item.enchantment;

import org.jetbrains.annotations.NotNull;

import java.lang.reflect.InvocationTargetException;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public abstract class Enchantment {

    private short level;

    public abstract short getId();

    public abstract int getMaxLevel();

    public short getLevel() {
        return this.level;
    }

    public @NotNull Enchantment setLevel(short level ) {
        this.level = level;
        return this;
    }

    public static <T extends Enchantment> @NotNull T create(EnchantmentType enchantmentType ) {
        try {
            return (T) EnchantmentRegistry.getEnchantmentClass( enchantmentType ).getConstructor().newInstance();
        } catch ( InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e ) {
            throw new RuntimeException( e );
        }
    }
}

File: JukeboxMC-API/src/main/kotlin/org/jukeboxmc/api/item/enchantment/Enchantment.kt
package org.jukeboxmc.api.item.enchantment

interface Enchantment {

    fun getId(): Int

    fun getLevel(): Int

    fun getMaxLevel(): Int

    fun getMinCost(level: Int): Int

    fun getMaxCost(level: Int): Int

    fun checkCompatibility(enchantment: Enchantment): Boolean

    fun getRarity(): Rarity

    override fun equals(other: Any?): Boolean

}